### ISSUE TITLE
Problems with the "acornadfs" format and Greaseweazle

### ISSUE THREAD

**Issue by Radiorama7** [davidgiven/fluxengine#340]
Hello,
just wanted to report the trouble I'm having with fluxengine+greaseweazle and the "acornadfs" format.

I'm using this command line to read Acorn ADFS disks with a Greaseweazle F7 Plus:
`fluxengine read acornadfs -o adfs.img --copy-flux-to=adfs.scp --decoder.write_csv_to=adfs.csv -s drive:1 --usb.greaseweazle=COM4
`

With some disks I get "multiple conflicting copies of sector xx seen" error messages, where xx is the track number; is it possible that fluxengine is mixing track numbers and sector numbers up?

With other disks I seem to get a correct read, like this:

Tracks -> 1         2         3         4         5         6         7
H.SS 01234567890123456789012345678901234567890123456789012345678901234567890123456789
0. 0 ................................................................................
0. 1 ................................................................................
0. 2 ................................................................................
.
.
0.15 ................................................................................
1. 0 ................................................................................
1. 1 ................................................................................
.
.
1.15 ................................................................................

but the image files I get are invariably 20 to 40 kB long instead of the expected 640 kB.

Also, the scp file is full of errors when decoded using HxC Floppy Emulator.

Reading an IBM PC 1.44 MB disk seems to work better: the image file is correct (as analyzed through HxC Floppy Emulator), but the flux file is still full of errors when decoded using HxC Floppy Emulator. Nevertheless, using the flux image file as the input to fluxengine (-s fluxfilename.scp instead of -s drive:1), a correct image file is still produced, therefore the scp file contains all the right data, and it is HxC Floppy Emulator that's unable to interpret the .scp file produced by fluxengine.

So to recap, my questions are:

Is the combination fluxengine+greaseweazle unable to read Acorn ADFS disks or am I doing something wrong?

Is fluxengine's .scp format different from what HxC expects?

Thanks!
Radiorama

**Comment by davidgiven** [comment]
Interesting. FluxEngine's SCP files should be reasonably standard, but there have been ongoing problems where the official SCP specification doesn't quite match what people are doing. Can you confirm that you're using the latest development build?

Regarding ADFS disks: you say 640kB, so that suggests a L format disk. The most common source of multiple definition errors is that you're seeing remnants of a different format previously on the disk, but L is 80-track double sided, so should be occupying the entire disk.

Can you attach the full console output of an attempt to read the disk, please? Also if you do `--write-flux-to=out.flux` and attach the flux file (assuming there's nothing sensitive there), I can take a look...

**Comment by Radiorama7** [comment]
Hi David, thanks for your prompt reply.

I'm using the latest build from 2021-10-31. The disk is the original Welcome disk for my Master Compact/PC 128 S to which I may have added a file or two when I was a kid, but surely nothing sensitive. It's an ADFS L disk.

I'm attaching a zip file with the .flux, .img, .csv files created by FluxEngine and a .txt with the console output.

What's the difference between a .scp and a .flux file? The latter seems to be smaller.

Thanks,
Radiorama

[adfs.zip](https://github.com/davidgiven/fluxengine/files/7455279/adfs.zip)

**Comment by davidgiven** [comment]
.flux files and .scp files are essentially the same thing; .flux is the native format used by FluxEngine, and .scp is the native format used by SuperCard Pro. FluxEngine's format is more flexible and stores more data, but is only readable by FluxEngine itself.

Having a look at the flux file it seems that the sectors on the two sides of the disk think they're on side 0. So, when the image is assembled, you get sector conflicts. I haven't actually read an ADFS L disk myself so it's possible they're supposed to look like that, but it's very weird.

You can tell FluxEngine to ignore the side byte in the sector IDs by adding `--decoder.ibm.ignore_side_byte=true`. This produces an actual image, but it's an Archimedes ADFS D or E file system containing a copy of "Time Detectives - The Victorians", so I think you used that disk for something else!

It's possible that `ignore_side_byte` needs to be set for ADFS disks. I need more information on the format.

**Comment by Radiorama7** [comment]
I made a .scp image of the Welcome disk using the Greaseweazle software, and I can confirm that the head byte is indeed always zero, even on side 1. I'm attaching the Greaseweazle image, that can be read and converted to other formats using the HxC software. I also see no trace of anything Archimedes on the disk, and it certainly boots as a Master Compact Welcome disk. The funny thing is that I never laid hands on an Archimedes, and the plot thickens further if we consider that the "Time Detectives" game you've mentioned is from 1994, while I got my Welcome disk with the computer back in 1988. Where on the disk did you find your references to "Time Detectives"?

I ran fluxengine again on the Welcome disk adding --decoder.ibm.ignore_side_byte=true as you suggested, and this time the disk image is 40 kB long instead of 20 kB. The files are in the attached welcome.zip.

I also found a games disk with the correct head number bytes and ran it as well; the results are in games.zip. No duplicate sectors this time, but the disk image is still 40 kB long.

Finally, returning to the Welcome disk without the ignore_side_byte as per the adfs.zip file attached to my previous message, there are exactly 16 "multiple conflicting copies of sector xx seen" messages for each xx from 0 to 79, as if the program were substituting track numbers for sector numbers.



[welcome.zip](https://github.com/davidgiven/fluxengine/files/7464674/welcome.zip)
[games.zip](https://github.com/davidgiven/fluxengine/files/7464679/games.zip)
[welcome-greaseweazle.zip](https://github.com/davidgiven/fluxengine/files/7464675/welcome-greaseweazle.zip)

**Comment by davidgiven** [comment]
Ah. Um. It's, er, possible --- _barely possible_ --- that I was looking at the wrong image file...

Okay, so after looking at the _right_ image, it seems that there's a subtle bug in the image writer where if you don't specify the number of sectors (which we can't because different ADFS disks use different numbers) but also specify the start sector ID, then the number of sectors would be assumed to be 1, I think. That should be fixed. I've also added a `ignore_track_byte` field because I initially thought that was what was happening, and changed the Acorn formats to include `ignore_side_byte`.

This is all in the `ibm` branch. PTAL?

Also note that FluxEngine produces interleaved image files, so you may not be able to use this directly with an emulator. (Traditionally Acorn emulators use file-system-block-order image files, where all of side 0 comes before all of side 1.)

**Comment by Radiorama7** [comment]
Thanks for looking into it. Where do I get the latest release?

**Comment by davidgiven** [comment]
I've merged the changes, so you should be able to get them from the latest development build at https://github.com/davidgiven/fluxengine/releases.

**Comment by Radiorama7** [comment]
Thank you very much - it works now!

**Comment by Radiorama7** [comment]
Although... is it possible that the graphical visualization obtained through fluxengine analyse layout ignores the index, ending up with sectors positioned at random angles?
![b82161c3-0e95-42b6-bf06-cc5a5428d599](https://user-images.githubusercontent.com/62107279/140435143-a1bc9f89-ed8e-415e-a72a-8cef797bf02c.jpg)

**Comment by davidgiven** [comment]
Actually, the _BBC Micro_ ignored the index pulse, and just wrote tracks whereever. BBC floppy drives tended not to even have a sensor for them. One of the things on my to-do list is to overhaul the visualiser to allow aligning tracks by a particular sector rather than the index pulse.

I'm glad this is working now!
